frontend/src/host-orchestrator/cmd/cvdserver-bootstrapper/main.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cvd-server-bootstrapper is a companion tool for the host orchestrator 
+// for bootstrapping AOSP cvd server.
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+)
+
+func hello(w http.ResponseWriter, req *http.Request) {
```

Comment:
why is an http server needed here?
This should just do the bootstrapping as soon as it starts. If this waits for an http request to come from elsewhere then that elsewhere will need to know when it's a good time to issue that request. For example, if the cloud orchestrator is the one making the request, it needs to know when this service is started and ready to run before issuing the request, which is not immediately after creating the host. Something else to take into account is that this port will need to be open in the firewall as well.

Reply:
the service won't be called by the cloud orchestrator, and the port does not have to be exposed, it only need to be called by the host-orchestrator.

I wanted to have a live service, cause I want the `cvd create` workflow to be aware of the `bootstrapping` stage rather than assume that there's going to be a cvd_server already up and running.

Having a live service managing the cvd_server has the extra benefit of updating the cvd_server itself which is a feature we could support once cvd_server can update itself.

Reply:
> the service won't be called by the cloud orchestrator, and the port does not have to be exposed, it only need to be called by the host-orchestrator.

You should probably use unix sockets instead of the network for communication between processes in the same host. You should be able to still use HTTP over unix sockets if you really want that.

> I wanted to have a live service, cause I want the cvd create workflow to be aware of the bootstrapping stage rather than assume that there's going to be a cvd_server already up and running.

But then you're assuming there is a cvdserver-bootstrapper already up and running, I don't think that's any different

> Having a live service managing the cvd_server has the extra benefit of updating the cvd_server itself which is a feature we could support once cvd_server can update itself.

That's actually a nice feature to have. I assume we'd ask to update the cvd server from the same place we ask to bootstrap it (likely the host-orchestrator), so why don't we wait until we decide to add that feature to add the http server and the request from the HO? The `cvd server` is still a work in progress, I'm worried they may change the way it's supposed to be updated and we end up doing this work for nothing.

Something else that I just remembered: The `cvd server` can not be correctly updated if there are instances running, so this may end up being harder than you think.

Reply:
great call, started using unix sockets instead.

> But then you're assuming there is a cvdserver-bootstrapper already up and running, I don't think that's any different

`cvdserver-bootstrapper` is installed together with `host-orchestrator` so it's a safer assumptions.

Reply:
I think there is an even simpler way to do this:

Put a `cvd` executable in the image that doesn't really contain the cvd executable, but instead contains the script that will download it and run it.

When the HO executes `cvd` the first time it will download the actual `cvd` binary from ci.android.com and execute that instead with whatever parameters HO gave it. If there is an error, it can communicate those back to the HO through stderr and/or the exit code. Further executions of `cvd` from the HO will call the actual `cvd` binary.

Problem 1: cvd server user
Solution: Use the setuid bit on both the bootstrapper and the cvd binary to ensure they always run as _cutf-executor. This is something you needed to do anyways in your http design to work around the case of the cvd server crashing and being restarted by an execution of the `cvd` binary from HO with the wrong user.

Problem 2: Updates
Solution: Don't delete the bootstrapper after downloading cvd, you can keep it in a different safe location and copy it over the cvd binary to force a new download the next time `cvd` is run or you can run the bootstrapper directly from that other location.

Reply:
tried the seuid bit idea, however it has an interesting drawback: the dynamic linker won't load shared libraries for cvd binaries spawned from cvd_server, i.e: cvd_internal_start, resulting in errors like:

```
/var/lib/cuttlefish-common/artifacts/8673413_aosp_cf_x86_64_phone-userdebug/bin/cvd_internal_start: error while loading shared libraries: libext2_blkid-host.so: cannot open shared object file: No such file or directory    
```

this is because the linker detects these binaries are being run in secure mode hence it does ignore environment variables like: LD_LIBRARY_PATH, see [ld.so.8](https://man7.org/linux/man-pages/man8/ld.so.8.html)